db/memtable.cc

```diff
@@ -583,54 +584,64 @@ bool MemTable::InsertKey(std::unique_ptr<MemTableRep>& table, KeyHandle handle,
   } else {
     iter->Seek(Slice(key_with_seq.data(), key_with_seq.size()));
   }
-  std::list<std::string> pinned_slices_;
+
   auto const& user_comparator = comparator_.comparator.user_comparator();
-  Slice last_start_key;
+  Slice last_key;
+  // start, end, tombstone seqnum
+  std::map<std::pair<Slice, Slice>,
+           std::set<SequenceNumber, std::greater<SequenceNumber>>,
+           decltype(cmp)>
+      to_insert(cmp);
+  // when we insert key@fragment_seq:value@tomstone_seqs, we ensure
+  // key@fragment_seq includes all the information in [key, value), so later
+  // insert only have to consider the entry key@fragment_seq and skip to later
+  // entry after [value, +inf)
+  bool done = false;
   for (; iter->Valid(); iter->Next()) {
     const Slice& ikey = iter->key();
     Slice tombstone_start_key = ExtractUserKey(ikey);
-    Slice tombstone_end_key = iter->value();
-    if (!last_start_key.empty() &&
-        user_comparator->Compare(last_start_key, tombstone_start_key) == 0) {
+    if (!last_key.empty() &&
+        user_comparator->Compare(tombstone_start_key, last_key) <= 0) {
```

Comment:
because last_key is now set last tombstone_end_key, and we only have to look at entry with tombstone_start_key >= last_key.

In our insert algorithm, We have ensured that if we insert entry B with tombstone_start_key < A.tombstone_end_key, we will insert an entry [B.tombstone_start_key, A.tombstone_end_key) and [A.tombstone_end_key,xxx).

when we have processed entry [B.tombstone_start_key, A.tombstone_end_key), assume A.tombstone_start_key > B.tombstone_start_key, we can ignore [A.tombstone_start_key, A.tombstone_end_key) and just process [A.tombstone_end_key, xxx)